### PR TITLE
[PHP 8.1 compatibility] Improve type checking for Transitions API method getTransitionsForAction 

### LINKED ISSUE #20590: Fatal error: htmlspecialchars_decode(): Argument #1 ($string) must be of type string, array given in Common.php
## Context
See below error details

> Error: {"message":"htmlspecialchars_decode(): Argument #1 ($string) must be of type string, array given","file":"core\/Common.php","line":422,"request_id":"3bc2b","backtrace":" on core\/Common.php(422)\n#0 core\/Common.php(422): htmlspecialchars_decode(Array, 3)\n#1 plugins\/Transitions\/API.php(165): Piwik\\Common::unsanitizeInputValue(Array)\n#2 plugins\/Transitions\/API.php(79): Piwik\\Plugins\\Transitions\\API->deriveIdAction(Array, 'url')\n#3 [internal function]: Piwik\\Plugins\\Transitions\\API->getTransitionsForAction(Array, 'url', '1', 'day', 'yesterday', false, 0, 'all')\n#4 core\/API\/Proxy.php(244): call_user_func_array(Array, Array)\n#5 core\/Context.php(28): Piwik\\API\\Proxy->Piwik\\API\\{closure}()\n#6 core\/API\/Proxy.php(335): Piwik\\Context::executeWithQueryParameters(Array, Object(Closure))\n#7 core\/API\/Request.php(267): Piwik\\API\\Proxy->call('\\\\Piwik\\\\Plugins\\\\...', 'getTransitionsF...', Array)\n#8 plugins\/API\/Controller.php(45): Piwik\\API\\Request->process()\n#9 [internal function]: Piwik\\Plugins\\API\\Controller->index()\n#10 core\/FrontController.php(637): call_user_func_array(Array, Array)\n#11 core\/FrontController.php(169): Piwik\\FrontController->doDispatch('API', false, Array)\n#12 core\/dispatch.php(32): Piwik\\FrontController->dispatch()\n#13 index.php(25): require_once('c...')\n#14 {main}","safemode_backtrace":"#0 [internal function]: Piwik\\Plugins\\Cloud\\Controller->safemode(Array)\n#1 \/core\/FrontController.php(637)

> URL: /index.php?module=API&method=Transitions.getTransitionsForAction&actionType=url&actionName[]=https%3A%2F%2Fwww.matomo.org%2Fde%2Fbar%2Ffoo%2F&idSite=1&period=day&date=yesterday&format=JSON&token_auth=XYZANONYMIZED&force_api_session=1

> Referrer: None

> GET: {"module":"API","method":"Transitions.getTransitionsForAction","actionType":"url","actionName":["https:\/\/www.matomo.org\/de\/bar\/foo\/"],"idSite":"1","period":"day","date":"yesterday","format":"JSON","token_auth":"XYZANONYMIZED","force_api_session":"1","filter_limit":100}

## Expected Behavior
No fatal error

## Current Behavior
Fatal error occurs

## Your Environment
Latest Matomo version

### PR DESCRIPTION
**Pull request by bx80**
### Description:

The transitions plugin `getTransitionsForAction()` API method doesn't handle the `actionName` parameter being passed as an array and throws a fatal error.

This PR allows a single element array or simple string to be correctly parsed, multi-element or empty arrays will return a NoDataForAction error.

Includes tests.

Fixes #20590

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

### CODE REVIEW

File: plugins/Transitions/API.php [hunk 1]
```diff
@@ -162,7 +162,10 @@ private function deriveIdAction($actionName, $actionType)
         switch ($actionType) {
             case 'url':
                 $originalActionName = $actionName;
-                $actionName = Common::unsanitizeInputValue($actionName);
+                if (is_array($actionName) && count($actionName) !== 1) {
+                    throw new Exception('NoDataForAction');
+                }
+                $actionName = (is_array($actionName) ? reset($actionName) : Common::unsanitizeInputValue($actionName));
```
Comment (michalkleiner) [inline plugins/Transitions/API.php]: Is this only specifically relevant to the `url` actionType or would it be a valid sanitisation for the `title` as well (or in general), and worth moving it before the switch statement?

Comment (bx80) [reply plugins/Transitions/API.php]: It would be a valid check for both `url` and `title` actions :+1:  I'll move it to the top of the method and add a few tests for `title` type actions too.

**Comment by sgiehl** [comment]
@bx80 We maybe should stop trying to support incorrectly provided parameters. The action name is meant to be passed as a string. We do not expect it to be an array. With Matomo 5 we introduced the possibility to typehint API parameters. So we could simply add a typehint to all of the API methods that expect action name or type to be a string. If a user provides it as an array the API should then give a proper error message.

**Comment by bx80** [comment]
Since we do have parameters elsewhere that do accept multiple types it's hard to know exactly what should be allowed where. Moving to type hints wherever possible will be a good way to solve this and add some certainty, though it might expose a few bugs on the way! 

I'll adjust this PR to simply add type hints and use that approach for any similar issues. :+1:

**Comment by sgiehl** [comment]
@bx80 we added a lot code in the past to handle incorrectly provided parameters as there was no easier way to do it and we wanted to avoid php errors to be thrown. But using the new typehints possibility makes it a lot simpler, as the API will directly return a proper error and we might actually be able to get rid of a lot error handling code in the APIs over time.

File: plugins/Transitions/API.php [hunk 2]
```diff
@@ -156,13 +156,19 @@ public function getTransitionsForAction($actionName, $actionType, $idSite, $peri
...
     /**
      * Derive the action ID from the request action name and type.
+     *
+     * @param string $actionName
+     * @param string $actionType
+     *
+     * @return array|int|string
      */
-    private function deriveIdAction($actionName, $actionType)
+    private function deriveIdAction(string $actionName, string $actionType)
```
Comment (sgiehl) [inline plugins/Transitions/API.php:165]: you need to add that to the public api methods as well, as only those are automatically checked and handled...

Comment (bx80) [reply plugins/Transitions/API.php:165]: Thanks, clearly wasn't thinking there  :man_facepalming: :slightly_smiling_face:

File: plugins/Transitions/API.php [hunk 3]
```diff
@@ -60,7 +60,7 @@ public function getTransitionsForPageUrl($pageUrl, $idSite, $period, $date, $seg
      * @return array
      * @throws Exception
      */
-    public function getTransitionsForAction($actionName, $actionType, $idSite, $period, $date,
+    public function getTransitionsForAction(string $actionName, string $actionType, $idSite, $period, $date,
```
Comment (michalkleiner) [inline plugins/Transitions/API.php:63]: I know it might be opening a can of worms here so we probably won't do it at this stage but if we're adding strict types, should we add them to all params for the given method? It feels a bit like "we don't care about the other ones" whereas it's just "we're not doing it at the moment".

Could we perhaps try adding all of them and see how the tests go and remove if too much stuff is failing?

We won't get the opportunity to change these until 6.x otherwise, correct?

Comment (bx80) [reply plugins/Transitions/API.php:63]: That would be nice, but some parameters like `$idSite` can legitimately have string and numeric values, eg. `1` or `all`, so it can be problematic to apply type hints without some degree of certainty that the parameter should only accept a single  type. Given that this was a time-boxed quick fix outside of the current sprint I'm reluctant to start wielding the can opener on this particular issue. :worm::worm::worm::worm:

Since many API methods receive the same parameters (idSite, period, date) it would be good to standardize on the strict types (if any) we can use for these across the whole API, and work towards having strict type on all API methods, but this will need some planning and is probably a 6.x thing. :slightly_smiling_face:

Comment (sgiehl) [reply plugins/Transitions/API.php:63]: @michalkleiner We can add such typehints to our API methods at any point, unless it would cause some sort of breaking change. Typecasts in API methods were not possible before Matomo 5, as our API processor wasn't able to handle them. Before parameters were actually always provided as `string`, `array` or `null`. Currently the most API methods don't use type hints, as it's hard to say what might break when simply adding them anywhere. My approach would be to add them whenever we need to work on a certain API method or add a new one and the types can be clearly defined.
This will also help use to fix certain bug reports that come in over and over again, where a user provides an invalid value type and the API throws a PHP error, as the type is handled correctly.

Comment (sgiehl) [inline plugins/Transitions/API.php:63]: This should also be changed for `getTransitionsForPageTitle` and `getTransitionsForPageUrl`, to ensure they provide a string value for the first parameter
